worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/MapPartitionDataReader.java

```diff
@@ -255,11 +241,44 @@ public synchronized void sendData() {
     }
   }
 
-  private long getIndexRegionSize() {
+  public RequestMessage generateReadDataMessage(
+      long streamId, int subPartitionId, ByteBuf byteBuf) {
+    return new ReadData(streamId, byteBuf);
+  }
+
+  protected void sendDataInternal(RecyclableBuffer buffer) {
+    final RecyclableBuffer wrappedBuffer = buffer;
+    int readableBytes = wrappedBuffer.byteBuf.readableBytes();
+    if (logger.isDebugEnabled()) {
+      logger.debug("send data start: {}, {}", streamId, readableBytes);
+    }
+
+    RequestMessage readData =
+        generateReadDataMessage(streamId, wrappedBuffer.subPartitionId, wrappedBuffer.byteBuf);
+    associatedChannel
+        .writeAndFlush(readData)
+        .addListener(
+            (ChannelFutureListener)
+                future -> {
+                  try {
+                    if (!future.isSuccess()) {
+                      recycleOnError(future.cause());
+                    }
+                  } finally {
+                    logger.debug("send data end: {}, {}", streamId, readableBytes);
+                    wrappedBuffer.recycle();
+                    numInUseBuffers.decrementAndGet();
+                  }
+                });
+    int currentCredit = credits.decrementAndGet();
```

Comment:
If future is failed due to certain exception, does this need to decrement?

Reply:
If the future is not success, this reader will be release, the value  of this field(i.e. credits) is meaningless then.